third_party/move/move-compiler-v2/src/pipeline/dead_store_elimination.rs

```diff
@@ -1,57 +1,202 @@
+/// - nodes that are known to be dead stores are marked as `dead`.
+///
+/// Note that this graph only contains side-effect-free *definitions* of the form:
+/// - `Assign(dst, src)`
+/// - `Load(dst, constant)`
+/// This is a conservative over-approximation of side-effect-free definitions.
+/// The nodes representing only *uses* have no restrictions like for definitions.
+/// A node can represent both a definition and a use: such a node can only be of the
+/// form `Assign(dst, src)`, which follows from the above restrictions.
+/// As such, many code offsets in a function that do not meet the above criteria
+/// may not be present in this graph.
+///
+/// A side-effect-free definition can be removed safely if it is not alive later.
+struct DefUseGraph {
+    children: BTreeMap<CodeOffset, BTreeSet<CodeOffset>>,
+    parents: BTreeMap<CodeOffset, BTreeSet<CodeOffset>>,
+    dead: BTreeSet<CodeOffset>,
+}
+
+impl DefUseGraph {
+    /// Create a new `DefUseGraph` from the function `target`.
+    pub fn new(target: &FunctionTarget) -> Self {
+        let mut this = Self {
+            children: BTreeMap::new(),
+            parents: BTreeMap::new(),
+            dead: BTreeSet::new(),
+        };
+        this.populate_from(target);
+        this
+    }
+
+    /// Obtain the set of dead stores, i.e., code offsets which can be removed safely.
+    pub fn dead_stores(mut self) -> BTreeSet<CodeOffset> {
+        let mut dead = BTreeSet::new();
+        while let Some(offset) = self.remove_a_dead_node() {
+            dead.insert(offset);
+        }
+        dead
+    }
+
+    /// Populate an empty graph from the (restricted) definitions and uses in `target`.
+    fn populate_from(&mut self, target: &FunctionTarget) {
+        let code = target.get_bytecode();
+        let live_vars = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("live variable annotation is a prerequisite");
+        for (offset, instr) in code.iter().enumerate() {
+            use Bytecode::*;
+            match instr {
+                Assign(_, dst, src, _) if dst == src => {
+                    // self-assignment is always a dead store
+                    self.incorporate_definition(*dst, offset as CodeOffset, live_vars, true);
+                },
+                Assign(_, dst, ..) | Load(_, dst, _) => {
+                    self.incorporate_definition(*dst, offset as CodeOffset, live_vars, false);
+                },
+                _ => {},
+            }
+        }
+    }
+
+    /// Remove one dead node (arbitrary but deterministic) from the graph, if present.
+    fn remove_a_dead_node(&mut self) -> Option<CodeOffset> {
+        if let Some(node) = self.dead.pop_last() {
+            let parents = self.disconnect_from_parents(node);
+            let children = self.disconnect_from_children(node);
+            // Reconnect all parents to all children.
+            for parent in parents.iter() {
```

Comment:
This doesn't make sense to me.  a uses b uses c; b is dead, so after removing b we say that a uses c?  No.

Reply:
The way things were setup, `b` could only be a self assignment (in which case, because it is a redef, connecting parents to children makes sense), or `b` had no live children (in which case, since all dead stuff is removed, it would be okay).

But I felt from you point that this was not obvious, so I have rewritten the algorithm in stages where it is easier to explain invariants.

Reply:
oh, dead might not mean unused, it might mean a useless self-assignment.

Reply:
Yes, dead store is when you remove the store and it should not affect the observable semantics (excluding gas semantics) of the program.